channeldb/control_tower.go

```diff
@@ -401,6 +489,131 @@ type InFlightPayment struct {
 	Attempt *PaymentAttemptInfo
 }
 
+// SubscribePayment subscribes to updates for the payment with the given hash.
+// It returns a boolean indicating whether the payment is still in flight and a
+// channel that provides the final outcome of the payment.
+func (p *paymentControl) SubscribePayment(paymentHash lntypes.Hash) (
+	bool, chan PaymentEvent, error) {
+
+	// Create a channel with buffer size 1. For every payment there will be
+	// exactly one event sent.
+	c := make(chan PaymentEvent, 1)
+
+	// Lock the hash to prevent a race condition with subscribing and an
+	// update from router.
+	p.hashMutex.Lock(paymentHash)
+	defer p.hashMutex.Unlock(paymentHash)
+
+	var inFlight bool
+
+	err := p.db.View(func(tx *bbolt.Tx) error {
+		bucket, err := fetchPaymentBucket(tx, paymentHash)
+		if err != nil {
+			return err
+		}
+
+		var event PaymentEvent
+
+		status := fetchPaymentStatus(bucket)
+		switch status {
+
+		// Payment is currently in flight. Register this subscriber and
+		// return without writing a result to the channel yet.
+		case StatusInFlight:
+			p.subscribersMtx.Lock()
+
+			subscribers, ok := p.subscribers[paymentHash]
+			if !ok {
+				subscribers = []chan PaymentEvent{c}
+
+			} else {
+				subscribers = append(subscribers, c)
+			}
+			p.subscribers[paymentHash] = subscribers
+			p.subscribersMtx.Unlock()
+
+			inFlight = true
+			return nil
+
+		// Payment already succeeded. It is not necessary to register as
+		// a subscriber, because we can send the result on the channel
+		// immediately.
+		case StatusSucceeded:
+			event.Status = StatusSucceeded
+
+			preimageSlice := bucket.Get(paymentSettleInfoKey)
+			if preimageSlice == nil {
+				return errors.New("missing preimage")
+			}
+			var err error
+			event.Preimage, err = lntypes.MakePreimage(preimageSlice)
+			if err != nil {
+				return err
+			}
+
+			a, err := fetchPaymentAttempt(bucket)
+			event.Route = &a.Route
+
+		// Payment already failed. It is not necessary to register as a
+		// subscriber, because we can send the result on the channel
+		// immediately.
+		case StatusFailed:
+			failureReasonSlice := bucket.Get(paymentFailInfoKey)
+			if failureReasonSlice == nil {
+				return errors.New("missing failure reason")
+			}
+
+			if len(failureReasonSlice) != 1 {
+				return errors.New("invalid failure reason data")
+			}
+
+			event.Status = StatusFailed
+			event.FailureReason = FailureReason(
+				failureReasonSlice[0],
+			)
+
+		default:
+			return errors.New("unknown payment status")
+		}
+
+		// Write immediate result to the channel.
+		c <- event
+		close(c)
+
+		return nil
+	})
+	if err != nil {
+		return false, nil, err
+	}
+
+	return inFlight, c, nil
```

Comment:
why is it necessary to distinguish a inflight payment here? Don't we know that it is either not initiated (which returns an error), or a payment is in flight and will be notified eventually?

Reply:
Then on the rpc level, we would always first return an inflight, followed by a succeeded. Even if the payment succeed already a long time ago.

Reply:
Which wouldn't _reeeeally_ make a difference for the caller, would it?

Reply:
It is a matter of preference. Imo it isn't the cleanest thing to start every lookup of a (historical) payment with an inflight notification